src/ring_theory/dedekind_domain/ideal.lean

```diff
@@ -855,3 +855,165 @@ end
 end is_dedekind_domain
 
 end height_one_spectrum
+
+section chinese_remainder
+
+open ideal unique_factorization_monoid
+open_locale big_operators
+
+variables {R}
+
+lemma ring.dimension_le_one.prime_le_prime_iff_eq (h : ring.dimension_le_one R)
+  {P Q : ideal R} [hP : P.is_prime] [hQ : Q.is_prime] (hP0 : P ≠ ⊥) :
+  P ≤ Q ↔ P = Q :=
+⟨(h P hP0 hP).eq_of_le hQ.ne_top, eq.le⟩
+
+variables {R} [is_domain R]
+
+lemma ideal.coprime_of_no_prime_ge {R : Type*} [comm_ring R] {I J : ideal R}
+  (h : ∀ P, I ≤ P → J ≤ P → ¬ is_prime P) : I ⊔ J = ⊤ :=
+begin
+  by_contra hIJ,
+  obtain ⟨P, hP, hIJ⟩ := ideal.exists_le_maximal _ hIJ,
+  exact h P (le_trans le_sup_left hIJ) (le_trans le_sup_right hIJ) hP.is_prime
+end
+
+section dedekind_domain
+
+variables [is_dedekind_domain R]
+
+lemma ideal.is_prime.mul_mem_pow (I : ideal R) [hI : I.is_prime] {a b : R} {n : ℕ}
+  (h : a * b ∈ I^n) : a ∈ I ∨ b ∈ I^n :=
+begin
+  cases n, { simp },
+  by_cases hI0 : I = ⊥, { simpa [pow_succ, hI0] using h },
+  simp only [← submodule.span_singleton_le_iff_mem, ideal.submodule_span_eq, ← ideal.dvd_iff_le,
+    ← ideal.span_singleton_mul_span_singleton] at h ⊢,
+  by_cases ha : I ∣ span {a},
+  { exact or.inl ha },
+  rw mul_comm at h,
+  exact or.inr (prime.pow_dvd_of_dvd_mul_right ((ideal.prime_iff_is_prime hI0).mpr hI) _ ha h),
+end
+
+section
+
+open_locale classical
+
+lemma ideal.count_normalized_factors_eq {p x : ideal R} (hp0 : p ≠ ⊥) [hp : p.is_prime] {n : ℕ}
+  (hle : x ≤ p^n) (hlt : ¬ (x ≤ p^(n+1))) :
+  (normalized_factors x).count p = n :=
+count_normalized_factors_eq ((ideal.prime_iff_is_prime hp0).mpr hp).irreducible (normalize_eq _)
+  (ideal.dvd_iff_le.mpr hle) (mt ideal.le_of_dvd hlt)
+
+end
+
+lemma ideal.le_mul_of_no_prime_factors
+  {I J K : ideal R} (coprime : ∀ P, J ≤ P → K ≤ P → ¬ is_prime P) (hJ : I ≤ J) (hK : I ≤ K) :
+  I ≤ J * K :=
+begin
+  simp only [← ideal.dvd_iff_le] at coprime hJ hK ⊢,
+  by_cases hJ0 : J = 0,
+  { simpa only [hJ0, zero_mul] using hJ },
+  obtain ⟨I', rfl⟩ := hK,
+  rw mul_comm,
+  exact mul_dvd_mul_left K
+    (unique_factorization_monoid.dvd_of_dvd_mul_right_of_no_prime_factors hJ0
+      (λ P hPJ hPK, mt ideal.is_prime_of_prime (coprime P hPJ hPK))
+      hJ)
+end
+
+lemma ideal.le_of_pow_le_prime {I P : ideal R} [hP : P.is_prime] {n : ℕ} (h : I^n ≤ P) : I ≤ P :=
+begin
+  by_cases hP0 : P = ⊥,
+  { simp only [hP0, le_bot_iff] at ⊢ h,
+    exact pow_eq_zero h },
+  rw ← ideal.dvd_iff_le at ⊢ h,
+  exact ((ideal.prime_iff_is_prime hP0).mpr hP).dvd_of_dvd_pow h
+end
+
+lemma ideal.pow_le_prime_iff {I P : ideal R} [hP : P.is_prime] {n : ℕ} (hn : n ≠ 0) :
+  I^n ≤ P ↔ I ≤ P :=
+⟨ideal.le_of_pow_le_prime, λ h, trans (ideal.pow_le_self hn) h⟩
+
+lemma ideal.prod_le_prime {ι : Type*} {s : finset ι} {f : ι → ideal R} {P : ideal R}
```

Comment:
Maybe `P` can be explicit here? But I am not sure.

Reply:
I'll leave it implicit since it occurs on both sides of the `↔` below.